v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/uniformsplitter/transforms/RangeBoundaryDoFn.java

```diff
@@ -0,0 +1,134 @@
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.uniformsplitter.columnboundary.ColumnForBoundaryQuery;
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.uniformsplitter.columnboundary.ColumnForBoundaryQueryPreparedStatementSetter;
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.uniformsplitter.range.BoundaryExtractorFactory;
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.uniformsplitter.range.BoundaryTypeMapper;
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.uniformsplitter.range.Range;
+import com.google.common.collect.ImmutableList;
+import java.io.Serializable;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import javax.annotation.Nullable;
+import javax.sql.DataSource;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** DoFn to find boundary (min, max) for a column (optionally for a given parent range). */
+final class RangeBoundaryDoFn extends DoFn<ColumnForBoundaryQuery, Range> implements Serializable {
+
+  private static final Logger logger = LoggerFactory.getLogger(RangeBoundaryDoFn.class);
+  private final SerializableFunction<Void, DataSource> dataSourceProviderFn;
+
+  private final UniformSplitterDBAdapter dbAdapter;
+
+  private String tableName;
+  private ImmutableList<String> partitionColumns;
+
+  @Nullable private BoundaryTypeMapper boundaryTypeMapper;
+
+  @JsonIgnore private transient @Nullable DataSource dataSource;
+
+  RangeBoundaryDoFn(
+      SerializableFunction<Void, DataSource> dataSourceProviderFn,
+      UniformSplitterDBAdapter dbAdapter,
+      String tableName,
+      ImmutableList<String> partitionColumns,
+      BoundaryTypeMapper boundaryTypeMapper) {
+    this.dataSourceProviderFn = dataSourceProviderFn;
+    this.dbAdapter = dbAdapter;
+    this.tableName = tableName;
+    this.partitionColumns = partitionColumns;
+    this.dataSource = null;
+    this.boundaryTypeMapper = boundaryTypeMapper;
+  }
+
+  @Setup
+  public void setup() throws Exception {
+    dataSource = dataSourceProviderFn.apply(null);
+  }
+
+  private Connection acquireConnection() throws SQLException {
+    return checkStateNotNull(this.dataSource).getConnection();
+  }
+
+  /**
+   * DoFn to find boundary (min, max) for a column (optionally for a given parent range).
+   *
+   * @param input Details for the column and parent range for which a boundary is requested.
+   * @param out new Range with column boundary.
+   * @param c process context.
+   * @throws SQLException - since this is in the run time, beam will auto retry the exception.
+   */
+  @ProcessElement
+  public void processElement(
+      @Element ColumnForBoundaryQuery input, OutputReceiver<Range> out, ProcessContext c)
+      throws SQLException {
+    String boundaryQuery =
```

Comment:
It is unclear to me where the limits are to come, but please check that values lower than min and larger than max detected at startup are not excluded while reading.

Reply:
The limits are fetched using a query of the type `select min(col), max(col)...` returned by the dbAdapter.

```
but please check that values lower than min and larger than max detected at startup are not excluded while reading.
```
Currently they are excluded. Wouldn't CDC cover for newer rows?
This behavior is consistent with JDBCIO#readWithPartitions, as in it reads only the slice between what it initially got as min and max.

I thought of keeping it consistent as I was a bit worried that reading more than max might be unbounded as - in a batch job might never fully go to a success state, for a table with auto incrementing keys and insert heavy workload.

Let's have a small catchup to discuss this more.